### PR TITLE
Database migration for HMAC Key Records

### PR DESCRIPTION
**Pull request by nplasterer**
Part of https://github.com/xmtp/libxmtp/issues/422

Adds the database table for hmac key records
Matches the structure currently used for hmac keys in V2 https://github.com/xmtp/proto/blob/main/proto/keystore_api/v1/keystore.proto#L306-L317

### CODE REVIEW

File: xmtp_mls/src/storage/encrypted_store/hmac_key_record.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            thirty_day_periods_since_epoch
+        }
+    }
+}
+
+impl_store!(StoredHmacKeyRecord, hmac_key_records);
+
+impl DbConnection {
+    /// Returns all hmac_key_records for the given group_id
+    pub fn get_hmac_key_records(
```
Comment (mchenani) [inline xmtp_mls/src/storage/encrypted_store/hmac_key_record.rs:53]: suggestion: does `impl_fetch_list_with_key!` do the job?

File: xmtp_mls/src/storage/encrypted_store/hmac_key_record.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    ) -> Result<Vec<StoredHmacKeyRecord>, StorageError> {
+        Ok(self.raw_query(|conn| -> diesel::QueryResult<_> {
+            dsl::hmac_key_records
+                .filter(dsl::group_id.eq(group_id))
+                .load::<StoredHmacKeyRecord>(conn)
+        })?)
+    }
+
+    /// Insert hmac_key_records without replacing existing ones
+    pub fn insert_hmac_key_records(
```
Comment (mchenani) [inline xmtp_mls/src/storage/encrypted_store/hmac_key_record.rs:65]: suggestion: can we use `impl_fetch_list_with_key!`

**Comment by mchenani** [comment]
By mistake update the PR from the main branch, I hope that would be ok!

File: xmtp_mls/migrations/2024-12-04-194513_hmac-keys-records-table/up.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE "hmac_key_records"(
+    -- Group ID that the Hmac keys are associated with
+    "group_id" BLOB NOT NULL,
```
Comment (nplasterer) [inline xmtp_mls/migrations/2024-12-04-194513_hmac-keys-records-table/up.sql:3]: I did this based off of the hmac keys work in V2 but @codabrink mentioned we might be doing this based on inboxId now. So maybe this is irrelevant in V3?

**Comment by richardhuaaa** [comment]
Some high-level thoughts, based on our previous discussion about deriving the hmac key from a user secret rather than a group secret in v3. It might be worth looping @codabrink into this PR too:

1. Right now the schema stores hmac keys for each `thirty_day_periods_since_epoch`. However, I think we need to store the user's root key from which the hmac keys are derived, otherwise we can't derive keys for future epochs.
2. If we are storing the root key, then maybe we don't need to store the derived keys, because we can just derive them when we need them. Then we don't need to worry about storing new records or deleting old records as time progresses.
3. I think there's an opportunity here to generalize this beyond hmacs. In a nutshell, all of a user's installations need to synchronize a root key used to derive hmacs. We could envision using this same system for other things, for example synchronizing a nickname. I'm wondering if what we want is a generic table for user settings/preferences.

**Comment by nplasterer** [comment]
Closing out for now infavor of a more general database long term for these kinds of preferences similar to our consent records DB.

**Pull request closed without merging**